Multi domain ledger api server fixes

oliverse-da opened the pull request:
### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [X] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

Closes #3351
Closes #3225
Closes #3324
Closes #2975
Closes #3327

stefanobaghino-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/index/StandaloneIndexServer.scala:
```diff
@@ -38,38 +39,21 @@ object StandaloneIndexServer {
       authService: AuthService,
-      loggerFactory: NamedLoggerFactory): StandaloneIndexServer =
+      loggerFactory: NamedLoggerFactory,
+      engine: Engine = engineSharedAmongIndexServers, // allows sharing DAML engine with DAML-on-X participant
```
I'm starting to think that this class does a little too much.

gerolf-da replied:
Is the purpose of this to not have to compile packages twice?

oliverse-da replied:
@gerolf-da , yes indeed

stefanobaghino-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/HikariJdbcConnectionProvider.scala:
```diff
@@ -18,7 +18,7 @@ trait JdbcConnectionProvider extends AutoCloseable {
     * The block must not recursively call [[runSQL]], as this could result in a deadlock
     * waiting for a free connection from the same pool. */
-  def runSQL[T](block: Connection => T): T
+  def runSQL[T](description: String, block: Connection => T): T
```
This spreads the logging concern across two classes, I'm not sure we should do this.

oliverse-da replied:
Agreed this looks strange.

gerolf-da replied:
why do we need it here as well?

oliverse-da replied:
The original intent was for errors, i.e. [in the HikariJdbcConnectionProvider where errors are caught](https://github.com/digital-asset/daml/blob/master/ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/HikariJdbcConnectionProvider.scala#L87-L89) to log the sql description as well, not just regular sql trace [in the SqlExecutor](https://github.com/digital-asset/daml/blob/master/ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/util/SqlExecutor.scala#L43-L44). This helps in regular production in which trace is off, but errors are logged.

SamirTalwar commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/HikariJdbcConnectionProvider.scala:
```diff
@@ -18,7 +18,7 @@ trait JdbcConnectionProvider extends AutoCloseable {
     * The block must not recursively call [[runSQL]], as this could result in a deadlock
     * waiting for a free connection from the same pool. */
-  def runSQL[T](block: Connection => T): T
+  def runSQL[T](description: String, block: Connection => T): T
```
Perhaps `runSQL[T](description: String)(block: Connection => T): T`?

This way it can be called like this:

```scala
runSQL("some description") { connection =>
  ...
}
```

oliverse-da replied:
Interestingly that's the version that I had in our repo, but then I noticed that [SQLExecutor](https://github.com/digital-asset/daml/blob/master/ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/util/SqlExecutor.scala#L36) combined the parameters in one list, so I combined them here as well.

oliverse-da replied:
Removed the HikariJdbcConnectionProvider.description parameter as part of moving the error logging from HikariJdbcConnectionProvider to SqlExecutor

stefanobaghino-da commented:
> Thanks for the feedback. Since all the changes besides the description-logging ones seem to be acceptable, I will drop #3324 (at least for this PR).

The effect of that change is valuable, is it not possible to have the same kind of output without having to pass the description in twice?

oliverse-da commented:
> > Thanks for the feedback. Since all the changes besides the description-logging ones seem to be acceptable, I will drop #3324 (at least for this PR).
> 
> The effect of that change is valuable, is it not possible to have the same kind of output without having to pass the description in twice?

The way to do it would be to move the error logging from the HikariJdbcConnectionProvider to SqlExecutor by adding a second try/catch-NonFatal-throw block in the latter.

stefanobaghino-da commented:
> The way to do it would be to move the error logging from the HikariJdbcConnectionProvider to SqlExecutor by adding a second try/catch-NonFatal-throw block in the latter.
If you can do it in this PR we can probably pull it in as well.

oliverse-da commented:
> If you can do it in this PR we can probably pull it in as well.

I have it coded up, but I am assessing the risk of logging something misleading in case a NonFatal exception is thrown not by the actual sql execution but the tracing and time-update code in the same outer try-catch:

```
      try {

        // Actual execution
        val res = block

        val execNanos = System.nanoTime() - startExec
        extraLog.foreach(log =>
          logger.trace(s"$description: $log exec ${(execNanos / 1E6).toLong} ms"))
        execTimer.update(execNanos, TimeUnit.NANOSECONDS)
        execAllTimer.update(execNanos, TimeUnit.NANOSECONDS)

        promise.success(res)
      } catch {
        case NonFatal(e) =>
+          logger.error(s"$description: Got an exception while executing a SQL query. Rolled back the transaction.", e)
          promise.failure(e)
        case t: Throwable =>
          logger.error("got a fatal error!", t) //fatal errors don't make it for some reason to the setUncaughtExceptionHandler above
          throw t
      }
    })
    promise.future
```

In such cases even the existing code returns a failed future while the connection has not been rolled back.

oliverse-da commented:
@stefanobaghino-da - I moved the logged error as discussed and changed the code to ignore exceptions thrown by the metrics section with 4770ee21d15a19d0b831dacadc434decba66caa8

stefanobaghino-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/util/SqlExecutor.scala:
```diff
@@ -46,23 +46,29 @@ final class SqlExecutor(noOfThread: Int, loggerFactory: NamedLoggerFactory, mm:
+          throw t
+      }
 
+      // decouple metrics updating from sql execution above
```
Great, thanks, very good catch!

No pun intended. :smile:

stefanobaghino-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/util/SqlExecutor.scala:
```diff
@@ -46,23 +46,29 @@ final class SqlExecutor(noOfThread: Int, loggerFactory: NamedLoggerFactory, mm:
-
         // Actual execution
-        val res = block
+        promise.success(block)
```
I may be wrong here, but if `promise` is immutable then you'll have to return it from this expression and then call `.future` on that in the end.

SamirTalwar replied:
I believe promises are necessarily mutable; you return the `promise.future` immediately and then can complete it later.

SamirTalwar replied:
According to [the Scala documentation](https://docs.scala-lang.org/overviews/core/futures.html#promises), it's mutable by design.

oliverse-da replied:
Indeed

stefanobaghino-da commented:
Just re-run Windows. @oliverse-da If you want this to be merged before release you have to rebase.
